example/main.py

```diff
@@ -1,163 +1,142 @@
-    parser.add_argument('--dataset', type=str, default='MNIST', help='Dataset to train with (CIFAR or MNIST)')
-
-    args = parser.parse_args()
+def main(args):
 
     transform = transforms.Compose([
                 transforms.ToTensor(),
                 transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))
             ])
 
-    trainloader, testloader = get_dataset(args.dataset, transform)
-
-    net = Net(num_channels=1 if args.dataset=='MNIST' else 3)
+    trainloader, testloader = get_dataset(args, transform)
+    net = LeNet()
 
-    criterion = nn.CrossEntropyLoss()
-    print('initializing downpour listener')
-    downpour_optim = optim.SGD(net.parameters(), lr=args.lr)
-    optimizer = DownpourSGD(net.parameters(), lr=args.lr, freq=10, model=net, internal_optim=downpour_optim)
-    # optimizer = optim.SGD(net.parameters(), lr=0.01, momentum=0.0)
+    if args.distributed:
+        optimizer = DownpourSGD(net.parameters(), lr=args.lr, freq=args.freq, model=net)
+    else:
+        optimizer = optim.SGD(net.parameters(), lr=args.lr, momentum=0.0)
+    scheduler = optim.lr_scheduler.ReduceLROnPlateau(optimizer, patience=1, verbose=True, min_lr=1e-3)
 
+    # train
     net.train()
-    num_print = 20
+    if args.cuda:
+        net = net.cuda()
+
     for epoch in range(args.epochs):  # loop over the dataset multiple times
         running_loss = 0.0
         for i, data in enumerate(trainloader, 0):
             # get the inputs
             inputs, labels = data
 
+            if args.cuda:
+                inputs, labels = inputs.cuda(), labels.cuda()
+
             # zero the parameter gradients
             optimizer.zero_grad()
             # forward + backward + optimize
             outputs = net(inputs)
-            loss = criterion(outputs, labels)
+            loss = F.cross_entropy(outputs, labels)
             loss.backward()
             optimizer.step()
 
             # print statistics
             running_loss += loss.item()
-            if i % num_print == 0:    # print every n mini-batches
-                print('Epoch: %d, Iteration: %5d loss: %.3f' % (epoch, i, running_loss / num_print))
+            if i % args.log_interval == 0 and i > 0:    # print every n mini-batches
+                print('Epoch: %d, Iteration: %5d loss: %.3f' % (epoch, i, running_loss / args.log_interval))
                 running_loss = 0.0
-        evaluate(net, testloader, args.dataset)
 
-    # join the listener
-    # optimizer.stop_listening()
+        val_loss = evaluate(net, testloader, args)
+        scheduler.step(val_loss)
+
     print('Finished Training')
 
 
-def evaluate(net, testloader, dataset_name):
-    if dataset_name == 'MNIST':
-        classes = (0, 1, 2, 3, 4, 5, 6, 7, 8, 9)
-    else:
-        classes = ('plane', 'car', 'bird', 'cat', 'deer', 'dog', 'frog', 'horse', 'ship', 'truck')
+def evaluate(net, testloader, args):
+    classes = ('plane', 'car', 'bird', 'cat', 'deer', 'dog', 'frog', 'horse', 'ship', 'truck')
```

Comment:
We should have the correct class names here for the MNIST dataset if that is the dataset we are training with as well.

Reply:
good idea will udpate